Optimize Instance serialization

pkch opened the pull request:
When `Instance` has no arguments, just represent it as a string, instead of a dict. Note: this change impacts not only `Instance` methods but also whoever is reading json from disk, since it's no longer true that every object is represented by a `dict` with a `.classname` key. As a result, this commit had to slightly change `Type.deserialize()`.

As discussed on gitter.

gvanrossum commented on mypy/types.py:
```diff
@@ -51,7 +51,9 @@ def serialize(self) -> JsonDict:
-    def deserialize(cls, data: JsonDict) -> 'Type':
+    def deserialize(cls, data: Union[JsonDict, str]) -> 'Type':
+        if isinstance(data, str):
+            return Instance.deserialize(data)
```
Actually you can just inline the code you added to `Instance.deserialize()` here -- then you won't have to change the argument to the latter to a Union.

gvanrossum commented on mypy/types.py:
```diff
@@ -437,16 +439,26 @@ def accept(self, visitor: 'TypeVisitor[T]') -> T:
         return data
 
     @classmethod
+    def deserialize_ex(cls, data: Union[JsonDict, str]) -> 'Instance':
```
I see no reason to make this a separate function. You can just fold this into Instance.deserialize().

pkch replied:
Fixed.

pkch commented:
Not ready for merge yet.

I accidentally cheated with type check: `Type.deserialize(cls, data: JsonDict) -> Type` actually receives `str` data sometimes (*mypy* doesn't raise warning because this parameter is always passed as a value of some `JsonDict`, so its type is `Any`). 

I don't think this is acceptable, but changing signature to `Type.deseriailze(cls, data: Union[JsonDict, str]) -> Type` would require updating `deserialize` in all `Type` subclasses to same, which is obviously bad (the only subclass that actually may receive a `str` is `Instance`).

The only fix I found was to factor split the new `Type.deserialize()` into `deserialize_ex()` that accepts a union, and the original `deserialize()` with the original signature. Then I replaced all `Type.deserialize()` with `Type.deserialize_ex()`. It works, but it's no good if anyone wants to call `super().deserialize()` with a string. Luckily, nobody needs that (`FunctionLike` is the only class that calls `deserialize()` through `super()`), but is it too fragile of a solution?

I added this fix as a separate commit. I am very curious about a good solution because this problem is kinda representative of adding type checks to traditional python code.

gvanrossum commented:
Maybe a solution would be to make Type.deserialize() a function, with a
Union argument. The way it's currently done is a cheat in itself. Then
there could be a proper Type.deserialize() that only takes a JsonDict and
always raises, and only Instance.deserialize() would need to take a Union
argument.

Also the whole idea of type-checking JSON data is broken anyways, so I
don't want to lose too much sleep over it. (If we had the ability to define
JsonDict as a recursive type we'd find lots of violations -- and indeed you
can trigger all sorts of errors by manually editing the JSON data.)

pkch commented:
Yes, I just saw the #182 and that thread perfectly summarized the problems I saw when I tried to redefine `JsonDict` in this case: if it's defined too strictly, a lot of code won't pass without a bunch of `isinstance`; and if it's defined loosely like now, mypy won't catch many type violations. I agree the latter is a lesser problem.

gvanrossum commented on mypy/types.py:
```diff
@@ -539,7 +551,7 @@ def with_name(self, name: str) -> 'FunctionLike': pass
     @classmethod
     def deserialize(cls, data: JsonDict) -> 'FunctionLike':
-        return cast(FunctionLike, super().deserialize(data))
+        return cast(FunctionLike, deserialize_type(data))
```
Hm... I think this method should be deleted and inlined in the one call site (in FuncDef.deserialize()).

pkch replied:
Fixed

gvanrossum replied:
You inlined it but didn't delete the method here.

pkch replied:
I need to learn to save modified files =( Fixed.

pkch commented:
The `touch mypy1/mypy/main.py; mypy -i mypy1/mypy` with warm cache fell from 1.04-1.10 sec to 0.98-1.00 sec on my laptop, so about 5-10%.

I also looked at cProfile output, but because in a few cases one method now does what several methods did previously, the change isn't too easy to analyze.

gvanrossum commented:
Great!

I hope that next you can look into reducing the bulk of serialized functions.
